packages/flutter/lib/src/widgets/shared_app_data.dart

```diff
@@ -0,0 +1,201 @@
+    final _SharedAppModel? model = InheritedModel.inheritFrom<_SharedAppModel>(context, aspect: key);
+    assert(_debugHasSharedAppData(model, context, 'getValue'));
+    return model!.sharedAppDataState.getValue<K, V>(key, init);
+  }
+
+  /// Changes the app model's `value` for `key` and rebuilds any widgets
+  /// that have created a dependency on `key` with [SharedAppData.getValue].
+  ///
+  /// If `value` is `==` to the current value of `key` then nothing
+  /// is rebuilt.
+  ///
+  /// The `value` is expected to be immutable because intrinsic
+  /// changes to the value will not cause dependent widgets to be
+  /// rebuilt.
+  ///
+  /// Unlike [SharedAppData.getValue], this method does _not_ create a dependency
+  /// between `context` and `key`.
+  ///
+  /// The type parameter `K` is the type of the value's keyword and `V`
+  /// is the type of the value.
+  static void setValue<K extends Object, V>(BuildContext context, K key, V value) {
+    final _SharedAppModel? model = context.getElementForInheritedWidgetOfExactType<_SharedAppModel>()?.widget as _SharedAppModel?;
+    assert(_debugHasSharedAppData(model, context, 'setValue'));
+    model!.sharedAppDataState.setValue<K, V>(key, value);
+  }
+
+  static bool _debugHasSharedAppData(_SharedAppModel? model, BuildContext context, String methodName) {
+    assert(() {
+      if (model == null) {
+        throw FlutterError.fromParts(
+          <DiagnosticsNode>[
+            ErrorSummary('No SharedAppData widget found.'),
+            ErrorDescription('SharedAppData.$methodName requires an SharedAppData widget ancestor.\n'),
+            context.describeWidget('The specific widget that could not find an SharedAppData ancestor was'),
+            context.describeOwnershipChain('The ownership chain for the affected widget is'),
+            ErrorHint(
+              'Typically, the SharedAppData widget is introduced by the MaterialApp '
+              'or WidgetsApp widget at the top of your application widget tree. It '
+              'provides a key/value map of data that is shared with the entire '
+              'application.',
+            ),
+          ],
+        );
+      }
+      return true;
+    }());
+    return true;
+  }
+}
+
+class _SharedAppDataState extends State<SharedAppData> {
+  late Map<Object, Object?> data = <Object, Object?>{};
+
+  @override
+  Widget build(BuildContext context) {
+    return _SharedAppModel(sharedAppDataState: this, child: widget.child);
+  }
+
+  V getValue<K extends Object, V>(K key, SharedAppDataInitCallback<V> init) {
+    data[key] ??= init();
+    return data[key] as V;
+  }
+
+  void setValue<K extends Object, V>(K key, V value) {
+    if (data[key] != value) {
+      setState(() {
+        data = Map<Object, Object?>.from(data);
+        data[key] = value;
+      });
+    }
+  }
+}
+
+class _SharedAppModel extends InheritedModel<Object> {
```

Comment:
Same here.

(It would probably be good to standardize on Model vs. Data to avoid confusion?)

Reply:
In this case I felt that the "Model" suffix was warranted since it's an InheritedModel